src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -1117,6 +1140,122 @@ int RadosStore::read_topics(const std::string& tenant, rgw_pubsub_topics& topics
   return 0;
 }
 
+int RadosStore::read_topics_all_tenants(std::map<std::string, rgw_pubsub_topics>& tenant_topics_map,
+        std::map<std::string, rgw_pubsub_bucket_topics>& tenant_bucket_notifs_map, optional_yield y, const DoutPrefixProvider *dpp) {
+  std::vector<std::string> raw_oids;
+  int ret = rgw_list_system_obj_filter_prefix(svc()->sysobj, svc()->zone->get_zone_params().log_pool,
+        pubsub_oid_prefix.substr(0, pubsub_oid_prefix.size()-1), raw_oids, dpp);
+  if (ret < 0) {
+    return ret;
+  }
+
+  // if contains bucket_notif_oid_infix  it is not a topic!!
+  for (const auto& oid: raw_oids) {
+    ldpp_dout(dpp, 20) << "read_topics_all_tenants oid: " << oid << dendl;
+    if (oid.find(bucket_notif_oid_infix ) != std::string::npos) {
+      rgw_pubsub_topics topics;
+      ret = read_topics(oid, topics, nullptr, y, dpp);
+      if (ret >= 0) {
+        ldpp_dout(dpp, 1) << "ERROR: failed to read topics on oid: " << oid << dendl;
+        tenant_topics_map.insert(std::make_pair(oid, topics));
+      }
+    } else {
+      rgw_pubsub_bucket_topics notifications;
+      std::string tenant, bucket_name, marker;
+      ret = RadosStore::deconstruct_topics_oid(oid, tenant, bucket_name, marker, dpp);
+      if (ret < 0) {
+        continue;
+      }
+      rgw_bucket rgw_bucket_info(tenant, bucket_name);
+      RadosBucket rados_bucket(this, rgw_bucket_info);
+
+      ret = rados_bucket.read_topics(notifications, nullptr, y, dpp);
+      if (ret >= 0) {
+        ldpp_dout(dpp, 1) << "ERROR: failed to read bucket notifications on oid: " << oid << dendl;
+        tenant_bucket_notifs_map.insert(std::make_pair(oid, notifications));
+      }
+    }
+  }
+
+  return 0;
+}
+
+int RadosStore::migrate_to_notification_data_path_v2(std::map<std::string, rgw_pubsub_topics>& tenant_topics_map,
+        std::map<std::string, rgw_pubsub_bucket_topics>& tenant_bucket_notifs_map, const DoutPrefixProvider *dpp, optional_yield y)
+{
+  ldpp_dout(dpp, 20) << "Info: starting migration for realm: " << svc()->site->get_realm()->name << dendl;
+  int ret = 0;
+  for (auto& tenant_topics_pair: tenant_topics_map) {
+    // start migration for topics
+    auto& tenant = tenant_topics_pair.first;
+    auto& topics = tenant_topics_pair.second;
+    RGWPubSub ps(this, tenant, *(svc()->site));
```

Comment:
please call directly into RadosStore for operations. RGWPubSub has extra layers of abstraction (the v1/v2 formats, and the sal::Driver backend) that we don't need here

the use of `RGWObjVersionTracker` and `bool exclusive` for handling races are really important here, but those aren't exposed by the `RGWPubSub` interface

https://github.com/ceph/ceph/pull/55504 has a mostly-complete example of this part in `migrate_topics()`

Reply:
Ack!